wpe/src/main/cpp/Browser/Page.cpp

```diff
@@ -1,15 +1,196 @@
 #include "Page.h"
 
 #include "Browser.h"
+#include "JNIHelper.h"
 #include "Logging.h"
 #include "RendererASurfaceTransaction.h"
 #include "RendererFallback.h"
 
 #include <algorithm>
 #include <android/api-level.h>
 #include <android/hardware_buffer.h>
+#include <android/native_window_jni.h>
+#include <jni.h>
 
 namespace {
+
+/** JNI Interface **/
+
+struct FieldIds {
+    jfieldID nativePtr;
+};
+FieldIds fields;
+
+inline Page* getPage(JNIEnv* env, jobject obj)
+{
+    return reinterpret_cast<Page*>(env->GetLongField(obj, fields.nativePtr));
+}
+
+void pageInit(JNIEnv* env, jobject obj, jint pageId, jint width, jint height)
+{
+    ALOGV("pageInit(%p, %d, %d, %d) [tid %d]", obj, pageId, width, height, gettid());
+    jclass clazz = env->GetObjectClass(obj);
+    jweak pageJavaRef = env->NewWeakGlobalRef(obj);
```

Comment:
I don't understand why storing this ref here.
It seems you are not using it anywhere as each time you are using the localRef of the jobject passed to the method (which is always valid because the method is called from the Java side).

I think that you may remove it.

Reply:
removed